docs/cplusplus-style-guide.md

```diff
@@ -0,0 +1,585 @@
+# C++ code style guidelines
+
+We use C++14, but currently consider C++17 too immature for usage.
+
+This documents contains guidelines regarding:
+
+* File structure
+* Comments and documentation
+* Tests
+* Naming
+* Formatting
+* Typing
+* Namespaces
+* Structs
+* Classes
+* Error handling
+* Memory handling
+* Resource handling
+* Other
+
+
+## File structure
+
+* Header files: `.hpp`
+* Source files: `.cpp`
+* Use `#pragma once` in headers.
+* Split class definition across a header and source file by default.
+* Include header definitions in the same file.
+
+* Order of includes
+	1. Related header.
+	2. `C` library.
+	3. `C++` library.
+	4. Other libraries.
+	5. Fetch library.
+
+* Classes in this order
+	1. `public`.
+	2. `protected`.
+	3. `private`.
+
+
+## Comments and documentation
+
+* Provide API documentation with your code.
+* Include a comment summary at the top of all files.
+
+``` c++
+/* Approximation of the exponential function.
+   * @x is the argument that is used in the exponent 2^(x / M_LN2).
+   *
+   * We exploit the IEEE-754 standard to efficiently create a linear
+   * interpolation between points of the form 2^n where n is an integer.
+   *
+   * You can find the full details in
+   * [1] "A fast, compact approximation of the exponential function",
+   * Nicol N. Schraudolph, 1999 Neural computation 11, 853-862.
+   *
+   * @return the exponential of x.
+   */
+double Exp(double const &x) 
+{
+    // We use the constants explained in [1] and additional make e the
+    // base rather than using 2.
+    static double const a = double(1ull << 20) / M_LN2;
+    static double const b = ((1ull << (10)) - 1) * double(1ull << 20); - 60801;
+    double in = x * a + b;
+
+    // Move the computed bits into the exponent with spilling into the
+    // Mantissa. The spilling creates the linear interpolation, while
+    // the exponent creates exact results whenever the argument is an integer.     
+    union 
+    { 
+      double d;
+      uint32_t i[2];
+    } conv;
+
+    conv.i[1] = static_cast<uint32_t>(in);
+    conv.i[0] = 0;
+
+    return conv.d;
+}
+```
+
+* Comment the non-obvious, not the obvious.
+
+``` c++
+// Turns out sorting this list before processing it is faster due to branch 
+// prediction, so don't remove this without checking the profile:
+std::sort(users.begin(), users.end(), UserSortPredicate());
+for (const User& user: users)
+{
+    if (user.ready)
+    {
+        // Do something...
+        
+    }
+}
+```
+
+* If you do not understand some code, contact the original author. If they have left the company ask the most competent person. After working out what a piece of a non-obvious code is doing, remember to document it.
+
+
+## Tests
+
+* Write tests and run them
+* Do not assume, do assert
+
+
+## Naming
+
+* Use meaningful names, even in tests.
+* For structs, use all lowercase with underscore to separate words, or camel case, i.e. `first_name` and `firstName` are both valid.
```

Comment:
This is not consistent. We use `Pascal` case for struct names. Attributes are `snake_case`

Reply:
Does the same apply to class names? I will document both for the sake of clarity.

Reply:
Yes